src/main/java/com/google/sps/data/Keywords.java

```diff
@@ -0,0 +1,132 @@
+package com.google.sps.data;
+
+import java.io.IOException;
+import com.google.cloud.language.v1.Document;
+import com.google.cloud.language.v1.LanguageServiceClient;
+import com.google.cloud.language.v1.AnalyzeEntitiesRequest;
+import com.google.cloud.language.v1.AnalyzeEntitiesResponse;
+import com.google.cloud.language.v1.EncodingType;
+import com.google.cloud.vision.v1.EntityAnnotation;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.TreeSet;
+
+/**
+ * Provide methods for the storage, extraction, and retrieval of keywords.
+ */
+public final class Keywords {
+
+  private static final int MAX_NUM_KEYWORDS = 10;
+
+  /**
+   * @return a list of the 10 most salient keywords
+   */
+  public static List<String> getKeywords(String key) throws EntityNotFoundException {
+    try {
+      Key datastoreKey = KeyFactory.stringToKey(key);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      Entity result = datastore.get(datastoreKey);
+    
+      Collection<String> keywordCollection = (Collection<String>) result.getProperty("keywords");
+      List<String> keywordList = new ArrayList<>();
+      int index = 0;
+      for (String keyword : keywordCollection) {
+        keywordList.add(keyword);
+    
+        index++;
+        if (index >= MAX_NUM_KEYWORDS) {
+          break;
+        }
+      }
+      return keywordList;
+    } catch (EntityNotFoundException ex) {
+      return new ArrayList<>();
+    }
+  }
+
+  /**
+   * Adds the labels from the analysis of an image into the datastore.
+   * @return the key pointing to the newly added datastore entity.
+   */
+  public static String addKeywords(List<EntityAnnotation> blobAnalysis) throws IOException {
+    String labels = "";
+    for (EntityAnnotation label : blobAnalysis) {
+      labels += label.getDescription() + ", ";
+    }
+    return addToDatastore(labels);
+  }
+
+  /**
+   * Adds the salient keywords from the analysis of a textual user input into the datastore.
+   * @return the key pointing to the newly added datastore entity.
+   */
+  public static String addKeywords(String message) throws IOException {
```

Comment:
I think I said this in an earlier PR, but do you need to make an extra function that's just calling another function.

Reply:
Yea, it helps streamline the function calls so I'm not calling addToDatastore for one thing and addKeywords for another.

Reply:
Yea, it helps streamline the function calls so I'm not calling addToDatastore for one thing and addKeywords for another.

Reply:
I can see the idea behind this but if that is the case I would recommend adding a comment explaining the reason for this function overload.  Otherwise, I think anyone who looks at this code would wonder why you are adding an extraneous layer of indirection.

Reply:
Thinking more about this I wonder if there is a better design.  It doesn't have to be done in this PR but it might be a good idea to use a more object orient style here.

Typically, when you need to overload methods to handle different classes, you could do things better by creating class methods.

e.g. Suppose I have the following code

```
public class Dog {
  private int age;
}

public class Cat {
  private float height;
}

public class Runner {
  private static void takeCareOfAnimals(List<Dog> dogs, List<Cat> cats) {
    for (Dog dog: dogs) {
      walk(dog)
    }
    for (Cat cat: cat) {
      walk(cat)
    }
  }

  private static void walk(Dog dog) {
    TimeUnit.MINUTES.sleep(5);
    system.out.println("Woof Woof");  
    TimeUnit.MINUTES.sleep(5);
  }

  private static void walk(Dog dog) {
    TimeUnit.MINUTES.sleep(5);
    system.out.println("Meow");  
    TimeUnit.MINUTES.sleep(5);
  }
}
```
The whole purpose of subclassing is to allow related objects to share properties without having to overload methods.  A better way to write the code would be as follows
```
public abstract class Animal {
  private abstract void MakeNoise();

  private void Walk() {
    TimeUnit.MINUTES.sleep(5);
    this.MakeNoise();  
    TimeUnit.MINUTES.sleep(5);      
  }
}

public class Dog extends Animal {
  private int age;

  private void MakeNoise() {
    system.out.println("Woof Woof"); 
  }
}

public class Cat {
  private float height;

  private void MakeNoise() {
    system.out.println("Meow"); 
  }
}

public class Runner {
  private static void takeCareOfAnimals(List<? extends Animal> animals) {
    for (Animal animal: animals) {
      walk(animal)
    }
  }
}
```
If the classes were a bit less related I might use an Interface (e.g. Walkable) instead of an AbstractClass but the same  principle holds.

I think you might be able to use this style to better design the code here.